### PR TITLE
feat: rework pubsub source to support parallel read and at-least-once

### LINKED ISSUE #16572: allow `>1` paralellism for pubsub source to increase ingestion performance
more discussion can be found here: https://risingwave-labs.slack.com/archives/C0606NNR74P/p1714493567060129

In essence, it is hard to make pubsub source exactly-once and other products only provide at-least-once semantics too, so 1 parallelism only does not help anyways.

We should make it parallel. 

cc: @fuyufjh

### PR DESCRIPTION
**Pull request by xxchan**
Signed-off-by: xxchan <xxchan22f@gmail.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close #16572

### Behavior changes

- Ack on checkpoint for Pub/Sub source.
	- (message) change the offset column from `ts` to `ack_id`
	- (split) don't update start_offset (for `seek`)
	- (reader) don't ack on read
	- refactor `WaitCheckpointWorker` to support `update_task_on_chunk`: store all `ack_ids` before next checkpoint
- Add (back) `pubsub.parallelism` WITH option. (Previously called `split_count`, but I like `parallelism` better, because "split" is not a user-facing concept #14531)
  - [ ] future work: use default parallelism when not specified. Like Flink `SourceEnumerator` `enumContext.currentParallelism()`. We don't have such a context in source manager yet.
- Removed the restriction of `retain_acked_messages`.

Note: although there are huge changes, I think all are none breaking.

### Explanations
Previously, we ack immediately and rely on `retain_acked_messages` + `seek` to timestamp to replay messages. This is awkward, and not like designed usage.

Now we ack on checkpoint and don't `seek` at all. It's stateless on RisingWave's side. This corresponds more the queueing semantics (delete after ack).

- See doc on `WaitCheckpointWorker`
- more questions discussed #16736
- previous reason for 1 parallelism https://github.com/risingwavelabs/risingwave/pull/14531#issuecomment-1892254000

### Testing
- case 1: Set `checkpoint_frequency` to very large (and `set visibility_mode TO 'all'`). Create Table and check load successfully. Kill and restart RisingWave, check the data can still be loaded without missing data (probably duplicated).
- case 2: Use normal config. Create Table and check load successfully without duplicated data.
- [x] (manual) test on emulator
	- [x] can we do recovery test in e2e test? https://github.com/risingwavelabs/risingwave/issues/16356
- [x] (manual) test on cloud

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Add `pubsub.parallelism` `WITH` option for PubSub source, which specifies the number of parallel consumers to run for the subscription. If not specified, the parallelism will be one.

### CODE REVIEW

File: src/connector/src/source/google_pubsub/mod.rs [hunk 1]
```diff
@@ -25,15 +28,19 @@ pub use source::*;
...
 
 #[derive(Clone, Debug, Deserialize, WithOptions)]
 pub struct PubsubProperties {
-    /// pubsub subscription to consume messages from
-    /// The subscription should be configured with the `retain-on-ack` property to enable
-    /// message recovery within risingwave.
+    /// Pub/Sub subscription to consume messages from.
+    ///
+    /// Note that we rely on Pub/Sub to load-balance messages between all Readers pulling from
+    /// the same subscription. So one `subscription` (i.e., one `Source`) can only used for one MV.
```
Comment (tabVersion) [inline src/connector/src/source/google_pubsub/mod.rs:41]: suggest: one subscription can be shared across one fragment to load-balance the parallelism.

File: src/connector/src/source/google_pubsub/mod.rs [hunk 2]
```diff
@@ -62,12 +69,17 @@ pub struct PubsubProperties {
...
-    /// Besides retention guarantees, timestamps are also more precise than timestamp-based seeks.
+    /// Besides retention guarantees, snapshots are also more precise than timestamp-based seeks.
     /// See [Seeking to a snapshot](https://cloud.google.com/pubsub/docs/replay-overview#seeking_to_a_timestamp) for
     /// more details.
     #[serde(rename = "pubsub.start_snapshot")]
     pub start_snapshot: Option<String>,
 
+    /// `parallelism` is the number of parallel consumers to run for the subscription.
+    /// TODO: use system parallelism if not set
+    #[serde(rename = "pubsub.parallelism")]
```
Comment (tabVersion) [inline src/connector/src/source/google_pubsub/mod.rs:80]:
```
#[serde(rename = "pubsub.parallelism")]
#[serde_as(as = "DisplayFromStr")]
pub parallelism: Option<u32>,
```

**Review by tabVersion — COMMENTED** [review]
the pr generally LGTM
please update in the pr's note that pubsub does not rely on the `SplitImpl`'s offset to start from a position. It prefers letting the broker side deal with offset, which is different from other connectors. 

btw, idk if reusing `offset` in the `SplitImpl` is a good practice, it may be confusing if users want to query from the state table.

File: src/connector/src/source/google_pubsub/mod.rs (hunk 2 above)
Comment (xxchan) [reply src/connector/src/source/google_pubsub/mod.rs:80]: IIRC `DisplayFromStr` doesn't work for `Option<T>`.

**Comment by xxchan** [comment]
> please update in the pr's note that pubsub does not rely on the `SplitImpl`'s offset to start from a position. It prefers letting the broker side deal with offset, which is different from other connectors.

Sure. BTW, I've mentioned that it will be **stateless** from the RisingWave side.

> btw, idk if reusing `offset` in the `SplitImpl` is a good practice, it may be confusing if users want to query from the state table.

I don't understand your point. What do you mean by reusing `offset`? Besides, actually `SplitImpl` doesn't maintain `offset` any more.

Anyway, I don't think any _average_ user care about query from state table.

File: src/connector/src/source/google_pubsub/split.rs [hunk 3]
```diff
@@ -29,10 +31,9 @@ pub struct PubsubSplit {
     /// pubsub reader.
     pub(crate) start_offset: Option<String>,
 
-    /// `stop_offset` is a numeric timestamp.
-    /// When not `None`, the `PubsubReader` stops reading messages when the `offset` property of
-    /// the `SourceMessage` is greater than or equal to the `stop_offset`.
-    pub(crate) stop_offset: Option<String>,
+    #[serde(rename = "stop_offset")]
+    #[serde(skip_serializing)]
+    pub(crate) __deprecated_stop_offset: Option<String>,
```
Comment (xxchan) [inline src/connector/src/source/google_pubsub/split.rs:36]: I guess `start_offset` can also be deprecated.

`stop_offset` is actually never used.

File: src/connector/src/source/google_pubsub/mod.rs (hunk 2 above)
Comment (tabVersion) [reply src/connector/src/source/google_pubsub/mod.rs:80]: Oh sorry, I mean `Option<DisplayFromStr>`

**Review by StrikeW — APPROVED** [review]
LGTM to `WaitCheckpointWorker`

**Comment by xxchan** [comment]
simple manual load testing

image

**Pull request merged**